# Data

In [81]:
from dotenv import load_dotenv
load_dotenv()

True

## 01 data loading

In [82]:
import os
import pandas as pd
import numpy as np

In [83]:
base_path = os.path.dirname(os.path.dirname(os.getcwd()))
x_mix_path = os.path.join(base_path,os.getenv("X_MIX_LABEL"))
y_mix_path = os.path.join(base_path,os.getenv("Y_LABEL"))

In [84]:
# 讀取 x 和 y 的資料
x_mix_gdp = pd.read_csv(x_mix_path)
y_mix_gdp = pd.read_csv(y_mix_path)

In [85]:
x_mix_gdp

年份  季別  平均匯率(元/美元)   國內投資毛額    投資率  工業生產指數  製造業銷售量指數  電力(企業)總用電量(十億度)  \
0    1982   1       37.91   102048  22.10   27.57     30.02             4.85   
1    1982   2       38.78   143589  29.41   30.35     33.54             5.46   
2    1982   3       39.65   117934  24.09   30.06     32.76             5.81   
3    1982   4       40.15   144192  28.92   29.86     32.70             5.66   
4    1983   1       39.97    99632  20.11   29.28     31.80             5.21   
..    ...  ..         ...      ...    ...     ...       ...              ...   
163  2022   4       31.37  1588290  26.94  220.26    211.99            32.25   
164  2023   1       30.40  1520462  27.70  188.21    187.55            28.45   
165  2023   2       30.71  1426025  24.97  188.72    195.97            31.84   
166  2023   3       31.69  1364758  22.69  205.88    207.63            34.72   
167  2023   4       31.84  1471478  23.06  215.34    209.11            32.09   

     進口物價基本分類指數  出口物價基本分類指數  ...  機械及電機設備進口值(十億元)  消費者物價指數   失業率  就業人數（千人）  \
0        231.63      311.21  ...             0.00   129.59  3.42  15826.00   
1        231.87      315.17  ...             0.00   130.53  4.12  15577.33   
2        234.50      317.22  ...             0.00   133.10  5.75  16000.33   
3        234.53      315.38  ...             0.00   132.11  6.41  16146.33   
4        227.62      309.54  ...             0.00   132.95  7.15  16260.67   
..          ...         ...  ...              ...      ...   ...       ...   
163      275.24      267.03  ...          1162.68   241.98  8.42  26657.00   
164      263.12      253.88  ...           925.86   243.54  8.22  26787.33   
165      257.68      252.57  ...           890.73   245.20  8.12  26852.33   
166      260.87      258.10  ...           973.80   246.62  8.28  26936.00   
167      263.75      261.96  ...          1049.92   249.07  7.88  26996.00   

     總人口數(千人)  景氣領先指標綜合指數(點)  景氣同時指標綜合指數(點)     gdp_t1     gdp_t2     gdp_t3  
0    42389.67          28.72          30.21        NaN        NaN        NaN  
1    42570.33          28.81          29.00   461774.0        NaN        NaN  
2    42742.00          28.50          28.87   488234.0   461774.0        NaN  
3    42952.67          29.05          29.58   489504.0   488234.0   461774.0  
4    43145.67          31.46          30.78   498511.0   489504.0   488234.0  
..        ...            ...            ...        ...        ...        ...  
163  53816.00         218.08         215.55  5718537.0  5558046.0  5647656.0  
164  54002.33         220.45         206.97  5896191.0  5718537.0  5558046.0  
165  54117.67         219.13         204.78  5489013.0  5896191.0  5718537.0  
166  54213.00         216.33         207.60  5710934.0  5489013.0  5896191.0  
167  54258.33         215.93         210.73  6016101.0  5710934.0  5489013.0  

[168 rows x 22 columns]

In [86]:
y_mix_gdp

年份  季別      GDP
0    1982   1   461774
1    1982   2   488234
2    1982   3   489504
3    1982   4   498511
4    1983   1   495361
..    ...  ..      ...
163  2022   4  5896191
164  2023   1  5489013
165  2023   2  5710934
166  2023   3  6016101
167  2023   4  6380686

[168 rows x 3 columns]

In [87]:
x_mix_gdp_drop=x_mix_gdp.drop(columns=['年份', '季別'])
y_mix_gdp_drop=y_mix_gdp.drop(columns=['年份', '季別'])

## 02 data MinMaxScaler

In [88]:
from sklearn.preprocessing import MinMaxScaler

In [89]:
scaler_X = MinMaxScaler()
scaler_y = MinMaxScaler()

T3_X_scaled = scaler_X.fit_transform(x_mix_gdp_drop)
T3_y_scaled = scaler_y.fit_transform(y_mix_gdp_drop)

## 03 Feaature with T3

- T3_X_scaled 
- T3_y_scaled

In [90]:
T3_x,T3_y=[],[]


for i in range(5,len(T3_X_scaled)):
    T3_x.append([T3_X_scaled[i]])

for i in range(5,len(T3_y_scaled)):
    T3_y.append([T3_y_scaled[i]])


T3_x=np.array(T3_x)
T3_x=T3_x.reshape(T3_x.shape[0], -1)

T3_y=np.array(T3_y)
T3_y=T3_y.reshape(T3_y.shape[0], -1)

In [91]:
T3_x.shape

(163, 20)

In [92]:
T3_y.shape

(163, 1)

## 04 data spliting 80 & 20

In [93]:
train_size = int(len(T3_x) * 0.8)

In [94]:
train_x_t3, test_x_t3 = T3_x[:train_size], T3_x[train_size:]
train_y_t3, test_y_t3 = T3_y[:train_size], T3_y[train_size:]

In [95]:
len(train_x_t3)

130

In [96]:
len(train_y_t3,)

130

In [97]:
len(test_x_t3)

33

In [98]:
len(test_y_t3)

33

In [99]:
# training data start in 1983 Q2
train_y_t3[0]

array([0.01459457])

In [100]:
# training data stop in 2015 Q3
train_y_t3[-1]

array([0.64621251])

In [101]:
# testing data start in 2015 Q4
test_y_t3[0]

array([0.66447448])

In [102]:
# testing data start in 2023 Q4
test_y_t3[-1]

array([1.])

In [103]:
real_train_y = scaler_y.inverse_transform(train_y_t3)
real_test_y = scaler_y.inverse_transform(test_y_t3)

# Model

- training data (train_y_t3 & train_x_t3) : 1983 q2 to 2015 q3 (130)
    - train_x_t3_scaled (x)

- testing data (test_y_t3 & test_x_t3) : 2015 q4 to 2023 q4 (33)
    - test_x_t3_scaled (x)


In [104]:
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [105]:
# 評估
def eval_metrics(y_true, y_pred, name=""):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_true, y_pred)
    print(f"{name} ➤ MSE: {mse:.2f}, RMSE: {rmse:.2f}, MAE: {mae:.2f}")
    return mse, rmse, mae

## GB

In [106]:
from sklearn.ensemble import GradientBoostingRegressor

In [107]:

gb_model = GradientBoostingRegressor(
    n_estimators=500,
    learning_rate=0.05,
    max_depth=5,
    min_samples_split=10,
    min_samples_leaf=5,
    subsample=0.9,
    max_features=0.8,
    random_state=42
)
gb_model.fit(train_x_t3, train_y_t3)


/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_gb.py:437: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingRegressor(learning_rate=0.05, max_depth=5, max_features=0.8,
                          min_samples_leaf=5, min_samples_split=10,
                          n_estimators=500, random_state=42, subsample=0.9)

In [108]:
# 預測
gb_y_train_pred = gb_model.predict(train_x_t3)
gb_y_test_pred = gb_model.predict(test_x_t3)


gb_y_train_pred = scaler_y.inverse_transform(gb_y_train_pred.reshape(-1, 1))
gb_y_test_pred = scaler_y.inverse_transform(gb_y_test_pred.reshape(-1, 1))

In [109]:
print("\n📊 訓練集表現：")
eval_metrics(real_train_y, gb_y_train_pred, "Train")

print("\n📊 測試集表現：")
eval_metrics(real_test_y, gb_y_test_pred, "Test")


📊 訓練集表現：
Train ➤ MSE: 152767.79, RMSE: 390.86, MAE: 276.94

📊 測試集表現：
Test ➤ MSE: 1076867615479.10, RMSE: 1037722.32, MAE: 816922.85


(1076867615479.1036, 1037722.3209891476, 816922.8532558331)

## Xgb

In [110]:
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split

In [111]:
xgb_model = XGBRegressor(
    n_estimators=500,
    learning_rate=0.05,
    max_depth=8,
    subsample=0.9,
    colsample_bytree=0.8,
    min_child_weight=5,
    gamma=0,
    random_state=42,
    reg_alpha=1,        
    reg_lambda=2,
)

xgb_model.fit(train_x_t3, train_y_t3)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.8, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=0, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.05, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=8,
             max_leaves=None, min_child_weight=5, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=500,
             n_jobs=None, num_parallel_tree=None, ...)

In [112]:
# 預測
xgb_y_train_pred = xgb_model.predict(train_x_t3)
xgb_y_test_pred = xgb_model.predict(test_x_t3)

xgb_y_train_pred = scaler_y.inverse_transform(xgb_y_train_pred.reshape(-1, 1))
xgb_y_test_pred = scaler_y.inverse_transform(xgb_y_test_pred.reshape(-1, 1))

In [113]:
print("\n📊 訓練集表現：")
eval_metrics(real_train_y, xgb_y_train_pred, "Train")

print("\n📊 測試集表現：")
eval_metrics(real_test_y, xgb_y_test_pred, "Test")


📊 訓練集表現：
Train ➤ MSE: 38392892329.77, RMSE: 195941.04, MAE: 104005.59

📊 測試集表現：
Test ➤ MSE: 2579115227716.12, RMSE: 1605962.40, MAE: 1497810.60


(2579115227716.123, 1605962.3992223863, 1497810.5984848484)

## RF

In [114]:
from sklearn.ensemble import RandomForestRegressor

In [115]:
rf_model = RandomForestRegressor(
    n_estimators=500, 
    min_samples_leaf=4,  
    max_features=0.8, 
    random_state=42,
)
rf_model.fit(train_x_t3, train_y_t3)

/var/folders/2n/bnnv_n796gg_v86gjzv7j07c0000gn/T/ipykernel_86949/3317303433.py:7: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model.fit(train_x_t3, train_y_t3)


RandomForestRegressor(max_features=0.8, min_samples_leaf=4, n_estimators=500,
                      random_state=42)

In [116]:
# 預測
rf_y_train_pred = rf_model.predict(train_x_t3)
rf_y_test_pred = rf_model.predict(test_x_t3)


rf_y_train_pred = scaler_y.inverse_transform(rf_y_train_pred.reshape(-1, 1))
rf_y_test_pred = scaler_y.inverse_transform(rf_y_test_pred.reshape(-1, 1))

In [117]:
print("\n📊 訓練集表現：")
eval_metrics(real_train_y, rf_y_train_pred, "Train")

print("\n📊 測試集表現：")
eval_metrics(real_test_y, rf_y_test_pred, "Test")


📊 訓練集表現：
Train ➤ MSE: 2155896954.92, RMSE: 46431.64, MAE: 32389.85

📊 測試集表現：
Test ➤ MSE: 1289088660624.54, RMSE: 1135380.40, MAE: 969244.63


(1289088660624.543, 1135380.4034879864, 969244.6272989165)